File: crates/aitrack-core/src/readers/claude.rs
//! Claude Code reader.

use crate::data::day_map::{add_model_usage, merge_day_maps, try_local_date_string};
use crate::data::model_id::strip_model_alias_suffix;
use crate::data::types::DayMap;
use crate::pricing::claude::estimate_claude_cost_usd;
use crate::pricing::fallback::FallbackCollector;
use crate::readers::cache::CachedParse;
use crate::readers::jsonl::stream_jsonl_objects;
use crate::readers::paths::get_claude_paths;
use crate::readers::pipeline::parse_provider_sources;
use serde_json::Value;
use std::collections::HashSet;
use std::path::{Path, PathBuf};

/// Parse one Claude transcript file against a dedup set.
fn parse_jsonl_file(
    file_path: &Path,
    seen: &mut HashSet<String>,
    fallbacks: Option<&mut FallbackCollector>,
) -> anyhow::Result<DayMap> {
    let mut result = DayMap::new();

    for obj in stream_jsonl_objects(file_path)? {
        // Check if type is "assistant"
        let entry_type = obj.get("type").and_then(|v| v.as_str());
        if entry_type != Some("assistant") {
            continue;
        }

        let message = match obj.get("message") {
            Some(Value::Object(m)) => m,
            _ => continue,
        };

        let usage = match message.get("usage") {
            Some(Value::Object(u)) => u,
            _ => continue,
        };

        let output_tokens = usage
            .get("output_tokens")
            .and_then(|v| v.as_i64())
            .unwrap_or(0);

        if output_tokens == 0 {
            continue;
        }

        // Build dedup key
        let message_id = message.get("id").and_then(|v| v.as_str()).unwrap_or("");
        let request_id = obj.get("requestId").and_then(|v| v.as_str()).unwrap_or("");
        let key = format!("{}:{}", message_id, request_id);

        if key != ":" && seen.contains(&key) {
            continue;
        }
        if key != ":" {
            seen.insert(key);
        }

        let timestamp = match obj.get("timestamp").and_then(|v| v.as_str()) {
            Some(ts) => ts,
            None => continue,
        };

        let date_string = match try_local_date_string(timestamp) {
            Some(ds) => ds,
            None => continue,
        };

        let model = strip_model_alias_suffix(
            message
                .get("model")
                .and_then(|v| v.as_str())
                .unwrap_or("unknown"),
        );

        let input_tokens = usage
            .get("input_tokens")
            .and_then(|v| v.as_u64())
            .unwrap_or(0);
        let cache_read = usage
            .get("cache_read_input_tokens")
            .and_then(|v| v.as_u64())
            .unwrap_or(0);
        let cache_creation = usage
            .get("cache_creation_input_tokens")
            .and_then(|v| v.as_u64())
            .unwrap_or(0);
        let output_tokens_u64 = output_tokens as u64;

        let total_input = input_tokens + cache_read + cache_creation;

        let claude_usage = crate::pricing::claude::ClaudeMessageUsage {
            input_tokens: Some(input_tokens),
            cache_read_input_tokens: Some(cache_read),
            output_tokens: Some(output_tokens_u64),
            cache_creation_input_tokens: Some(cache_creation),
        };

        let cost_usd = estimate_claude_cost_usd(
            &model,
            &claude_usage,
            Some(&date_string),
            fallbacks.as_deref(),
        );

        let day = result.entry(date_string).or_default();

        add_model_usage(
            day,
            &model,
            &crate::data::types::TokenCounts {
                input_tokens: total_input,
                output_tokens: output_tokens_u64,
                cached_input_tokens: Some(cache_read),
                raw_input_tokens: Some(input_tokens),
                cache_creation_input_tokens: Some(cache_creation),
                cost_usd: Some(cost_usd),
            },
        );
    }

    Ok(result)
}

/// Parse one Claude file in isolation (for caching).
pub fn parse_claude_file(file_path: &Path) -> anyhow::Result<CachedParse> {
    let mut keys = HashSet::new();
    let days = parse_jsonl_file(file_path, &mut keys, None)?;
    Ok(CachedParse {
        days,
        keys: keys.into_iter().collect(),
    })
}

/// Fold per-file parses into one DayMap, re-reading files with colliding keys.
async fn merge_claude_parsed(
    parsed: Vec<CachedParse>,
    files: Vec<PathBuf>,
) -> anyhow::Result<DayMap> {
    let mut all_days = DayMap::new();
    let mut seen_messages = HashSet::new();

    for (index, entry) in parsed.iter().enumerate() {
        let file_path = &files[index];

        // Check if any keys in this file were already seen
        if entry.keys.iter().any(|key| seen_messages.contains(key)) {
            // Re-read this file against the running set
            let fresh_days = parse_jsonl_file(file_path, &mut seen_messages, None)?;
            merge_day_maps(&mut all_days, &fresh_days);
            continue;
        }

        // No collisions, use cached data
        for key in &entry.keys {
            seen_messages.insert(key.clone());
        }
        merge_day_maps(&mut all_days, &entry.days);
    }

    Ok(all_days)
}

pub fn read_claude_data(_fallbacks: Option<&mut FallbackCollector>) -> anyhow::Result<DayMap> {
    let roots = get_claude_paths();
    let result = parse_provider_sources("claude", roots, parse_claude_file)?;

    // Use pollster instead of futures for blocking
    pollster::block_on(merge_claude_parsed(result.parsed, result.files))
}

File: packages/aitrack-lib/src/readers/claude.ts
import { homedir } from 'node:os';
import { join } from 'node:path';

import { tryLoadConfig } from '../config.js';
import { addModelUsage, getOrCreateDay, mergeDayMaps, tryLocalDateString } from '../data/dayMap.js';
import { stripModelAliasSuffix } from '../data/modelId.js';
import type { DayMap } from '../data/types.js';
import { environmentValue } from '../env.js';
import { claudeCacheWriteTokens, estimateClaudeCostUSD } from '../pricing/claude.js';
import type { FallbackCollector } from '../pricing/fallback.js';
import type { CachedParse } from './cache.js';
import { streamJsonlObjects } from './jsonl.js';
import { resolveSourceRoots } from './paths.js';
import { parseProviderSources } from './pipeline.js';

export function getClaudePaths(): string[] {
  const xdg = environmentValue('XDG_CONFIG_HOME');
  return resolveSourceRoots({
    envValue: environmentValue('AITRACK_CLAUDE_PROJECTS_DIRS'),
    configValue: tryLoadConfig()?.claudeProjectsDir,
    defaults: [
      ...(xdg ? [join(xdg, 'claude', 'projects')] : []),
      join(homedir(), '.config', 'claude', 'projects'),
      join(homedir(), '.claude', 'projects'),
    ],
  });
}

interface ClaudeEntry {
  type: string;
  timestamp?: string;
  requestId?: string;
  message?: {
    id?: string;
    model?: string;
    usage?: {
      input_tokens?: number;
      cache_read_input_tokens?: number;
      output_tokens?: number;
      cache_creation_input_tokens?: number;
      cache_creation?: {
        ephemeral_5m_input_tokens?: number;
        ephemeral_1h_input_tokens?: number;
      };
    };
  };
}

interface CountedMessage {
  dateString: string;
  model: string;
  inputTokens: number;
  outputTokens: number;
  rawInputTokens: number;
  cachedInputTokens: number;
  cacheCreationInputTokens: number;
  cacheCreation1hInputTokens: number;
  costUSD?: number;
}

/** Prefer more output, then more input. A later record wins a tie. */
function isRicherUsage(next: CountedMessage, current: CountedMessage): boolean {
  if (next.outputTokens !== current.outputTokens) return next.outputTokens > current.outputTokens;
  if (next.inputTokens !== current.inputTokens) return next.inputTokens > current.inputTokens;
  return true;
}

export async function parseJsonlFile(
  filePath: string,
  seen: Set<string>,
  fallbacks?: FallbackCollector,
): Promise<DayMap> {
  const result: DayMap = new Map();
  const chosen = new Map<string, CountedMessage>();
  const unkeyed: CountedMessage[] = [];

  for await (const parsed of streamJsonlObjects(filePath)) {
    const entry = parsed as unknown as ClaudeEntry;

    if (entry.type !== 'assistant') continue;
    const usage = entry.message?.usage;
    if (!usage) continue;

    const ts = entry.timestamp;
    if (!ts) continue;
    const dateString = tryLocalDateString(ts);
    if (dateString === null) continue;

    const writes = claudeCacheWriteTokens(usage);
    const inputTokens =
      (usage.input_tokens ?? 0) +
      (usage.cache_read_input_tokens ?? 0) +
      writes.fiveMinute +
      writes.oneHour;
    const outputTokens = usage.output_tokens ?? 0;
    if (inputTokens === 0 && outputTokens === 0) continue;

    const model = stripModelAliasSuffix(entry.message?.model ?? 'unknown');
    const costUSD = estimateClaudeCostUSD(model, usage, dateString, fallbacks);
    const counted: CountedMessage = {
      dateString,
      model,
      inputTokens,
      outputTokens,
      rawInputTokens: usage.input_tokens ?? 0,
      cachedInputTokens: usage.cache_read_input_tokens ?? 0,
      cacheCreationInputTokens: writes.fiveMinute,
      cacheCreation1hInputTokens: writes.oneHour,
      ...(costUSD !== undefined && { costUSD }),
    };

    // No id to dedupe on: count the row. A known id keeps the fullest record
    // in this file (a later correction replaces a partial stream entry). Keys
    // already counted by an earlier file stay skipped so a resumed transcript
    // is not added twice.
    const key = `${entry.message?.id ?? ''}:${entry.requestId ?? ''}`;
    if (key === ':') {
      unkeyed.push(counted);
      continue;
    }
    if (seen.has(key)) continue;
    const existing = chosen.get(key);
    if (!existing || isRicherUsage(counted, existing)) chosen.set(key, counted);
  }

  for (const [key, counted] of chosen) {
    seen.add(key);
    addCountedMessage(result, counted);
  }
  for (const counted of unkeyed) addCountedMessage(result, counted);

  return result;
}

function addCountedMessage(result: DayMap, counted: CountedMessage): void {
  const day = getOrCreateDay(result, counted.dateString);
  addModelUsage(day, counted.model, {
    inputTokens: counted.inputTokens,
    outputTokens: counted.outputTokens,
    rawInputTokens: counted.rawInputTokens,
    cachedInputTokens: counted.cachedInputTokens,
    cacheCreationInputTokens: counted.cacheCreationInputTokens,
    ...(counted.cacheCreation1hInputTokens > 0 && {
      cacheCreation1hInputTokens: counted.cacheCreation1hInputTokens,
    }),
    ...(counted.costUSD !== undefined && { costUSD: counted.costUSD }),
  });
}

/**
 * Parse one transcript in isolation, collecting the dedup keys it holds.
 *
 * The per-file view is what the cache stores: it depends only on the file's own
 * bytes, unlike a parse threaded through the corpus-wide `seen` set. Passing a
 * fresh set both de-duplicates within the file and leaves the file's keys
 * behind for the cross-file check in readClaudeData.
 */
export async function parseClaudeFile(
  filePath: string,
  fallbacks?: FallbackCollector,
): Promise<CachedParse> {
  const keys = new Set<string>();
  const days = await parseJsonlFile(filePath, keys, fallbacks);
  return { days, keys: [...keys] };
}

/**
 * Fold the per-file parses into one DayMap, re-reading any file whose messages
 * an earlier file already counted (a resumed session copies another
 * transcript's history into itself, which the per-file parse cannot see).
 */
export async function mergeClaudeParsed(parsed: CachedParse[], files: string[]): Promise<DayMap> {
  const allDays: DayMap = new Map();
  const seenMessages = new Set<string>();
  for (const [index, entry] of parsed.entries()) {
    const filePath = files[index];
    if (filePath !== undefined && entry.keys.some((key) => seenMessages.has(key))) {
      // Re-read this file against the running set, which is what an uncached
      // run would have done.
      mergeDayMaps(allDays, await parseJsonlFile(filePath, seenMessages));
      continue;
    }
    for (const key of entry.keys) seenMessages.add(key);
    mergeDayMaps(allDays, entry.days);
  }
  return allDays;
}

export async function readClaudeData(fallbacks?: FallbackCollector): Promise<DayMap> {
  const { files, parsed } = await parseProviderSources({
    cacheName: 'claude',
    roots: getClaudePaths(),
    parseFile: parseClaudeFile,
    fallbacks,
  });
  return mergeClaudeParsed(parsed, files);
}
